### PR TITLE
CADC-11395 Added recursive set node feature to cavern.

### PR DESCRIPTION
**Pull request by yeunga**

### CODE REVIEW

**Review by hjeeves — CHANGES_REQUESTED** [review]
Really minor things about date stamps on headers and commented code. :) Otherwise looks good.

File: cavern/src/intTest/java/org/opencadc/cavern/RecursiveSetNodeRunnerTest.java [hunk 1]
```diff
@@ -0,0 +1,595 @@
+/*
+************************************************************************
+*******************  CANADIAN ASTRONOMY DATA CENTRE  *******************
+**************  CENTRE CANADIEN DE DONNÉES ASTRONOMIQUES  **************
+*
+*  (c) 2011.                            (c) 2011.
```
[the remark this answers is not in the document]
Comment (yeunga) [reply cavern/src/intTest/java/org/opencadc/cavern/RecursiveSetNodeRunnerTest.java:6]: Updated

File: cavern/src/intTest/java/org/opencadc/cavern/RecursiveSetNodeRunnerTest.java [hunk 2]
```diff
@@ -0,0 +1,595 @@
...
+        Assert.assertTrue("expected Success count: 0 Failure count: 1", js.es.getSummaryMessage().equalsIgnoreCase("Success count: 0 Failure count: 1"));
+    }
+
+    @Test
+    public void testSingleNode() throws Throwable
+    {
+        VOSpaceClient vos = new VOSpaceClient(nodeURI.getServiceURI());
+        Subject s = SSLUtil.createSubject(SSL_CERT);
+        DataNode data = new DataNode(new VOSURI(new URI(nodeURI + "/dataNode-" + System.currentTimeMillis())));
+//        DataNode data = new DataNode(new VOSURI(new URI(nodeURI + "/dataNode")));
```
[the remark this answers is not in the document]
Comment (yeunga) [reply cavern/src/intTest/java/org/opencadc/cavern/RecursiveSetNodeRunnerTest.java:289]: Removed.

File: cavern/src/main/java/org/opencadc/cavern/JobAbortedException.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+package org.opencadc.cavern;
+
+import ca.nrc.cadc.uws.Job;
+import ca.nrc.cadc.vos.VOSException;
+
+/**
+ * Exception indicating the a job has been aborted by the client.
+ * 
+ * @author majorb
```
[the remark this answers is not in the document]
Comment (yeunga) [reply cavern/src/main/java/org/opencadc/cavern/JobAbortedException.java:78]: Updated.

File: cavern/src/main/java/org/opencadc/cavern/RecursiveSetNodeJobManager.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
+/*
+************************************************************************
+*******************  CANADIAN ASTRONOMY DATA CENTRE  *******************
+**************  CENTRE CANADIEN DE DONNÉES ASTRONOMIQUES  **************
+*
+*  (c) 2009.                            (c) 2009.
```
[the remark this answers is not in the document]
Comment (yeunga) [reply cavern/src/main/java/org/opencadc/cavern/RecursiveSetNodeJobManager.java:6]: Updated.

File: cavern/src/main/java/org/opencadc/cavern/RecursiveSetNodeJobManager.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+import org.apache.log4j.Logger;
+
+import ca.nrc.cadc.uws.server.JobExecutor;
+import ca.nrc.cadc.uws.server.JobUpdater;
+import ca.nrc.cadc.uws.server.ThreadPoolExecutor;
+
+/**
+ *
+ * @author pdowler, majorb
```
[the remark this answers is not in the document]
Comment (yeunga) [reply cavern/src/main/java/org/opencadc/cavern/RecursiveSetNodeJobManager.java:80]: Added my name to the list.

File: cavern/src/main/java/org/opencadc/cavern/RecursiveSetNodeRunner.java [hunk 6]
```diff
@@ -0,0 +1,377 @@
+/*
+************************************************************************
+*******************  CANADIAN ASTRONOMY DATA CENTRE  *******************
+**************  CENTRE CANADIEN DE DONNÉES ASTRONOMIQUES  **************
+*
+*  (c) 2009.                            (c) 2009.
```
[the remark this answers is not in the document]
Comment (yeunga) [reply cavern/src/main/java/org/opencadc/cavern/RecursiveSetNodeRunner.java:6]: Updated.

File: cavern/src/main/java/org/opencadc/cavern/RecursiveSetNodeRunner.java [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+import java.security.AccessControlException;
+import java.util.Date;
+import java.util.List;
+import org.apache.log4j.Logger;
+
+/**
+ * Class to set node properties on the target node and all child
+ * nodes recursively.
+ *
+ * @author majorb
```
[the remark this answers is not in the document]
Comment (yeunga) [reply cavern/src/main/java/org/opencadc/cavern/RecursiveSetNodeRunner.java:70]: Added my name as an author.

File: cavern/src/main/java/org/opencadc/cavern/RecursiveSetNodeRunner.java [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+    }
+
+    @Override
+    public void run()
+    {
+        log.debug("RUN RecursiveSetNodeRunner");
+        logInfo = new JobLogInfo(job);
+
+        String startMessage = logInfo.start();
+        log.info(startMessage);
```
[the remark this answers is not in the document]
Comment (yeunga) [reply cavern/src/main/java/org/opencadc/cavern/RecursiveSetNodeRunner.java:115]: I think timing statements are okay. Just following what vault is doing.

File: cavern/src/main/java/org/opencadc/cavern/RecursiveSetNodeRunner.java [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+                {
+                    // don't update the first child if this is not the
+                    // first batch request - it's been done already.
+                    if (lastChildURI == null || !child.getUri().equals(lastChildURI))
+                    {
+                        this.applyNodeProperties(child, properties);
+                    }
+                }
+
+                if (container.getNodes().size() == CHILD_BATCH_SIZE)
```
[the remark this answers is not in the document]
Comment (yeunga) [reply cavern/src/main/java/org/opencadc/cavern/RecursiveSetNodeRunner.java:303]: Corrected all braces.

File: cavern/src/main/java/org/opencadc/cavern/JobAbortedException.java [hunk 10]
```diff
@@ -0,0 +1,105 @@
+/*
+************************************************************************
+*******************  CANADIAN ASTRONOMY DATA CENTRE  *******************
+**************  CENTRE CANADIEN DE DONNÉES ASTRONOMIQUES  **************
+*
+*  (c) 2009.                            (c) 2009.
```
[the remark this answers is not in the document]
Comment (yeunga) [reply cavern/src/main/java/org/opencadc/cavern/JobAbortedException.java]: Updated to 2023.

**Comment by yeunga** [comment]
Looks like the threee check failures are the result of Pat's updates to A&A currently in progress.

**Review by brianmajor — APPROVED** [review]

**Pull request merged**